### PR TITLE
Add support for build modes 

### PR DESCRIPTION
**Pull request by aholstrup1**
Adding support for build modes. Initial support the following build modes: 
- **Translated** - Running a build with 'translationfile' as a compiler feature
- **Clean** - Applying a set of _preprocessorsymbols_ during compilation. The list of _preprocessorsymbols_ is defined in the settings.json of the AL-GO project 
- **Default** - Running a build with the code as it is.

Default build mode is 🥁...  **Default**

### CODE REVIEW

File: Actions/Deliver/Deliver.ps1 [hunk 1]
```diff
@@ -163,37 +168,42 @@ try {
...
-            if ($dependenciesFolder.Count -gt 1) {
-                $dependenciesFolder | Out-Host
-                throw "Internal error - multiple dependencies folders located"
-            }
-            elseif ($dependenciesFolder.Count -eq 1) {
-                $parameters.dependenciesFolder = $dependenciesFolder[0]
-            }
-            else {
-                $parameters.dependenciesFolder = ""
+            $projectToFilter = $project.Replace('\','_') # this is the project name as it appears in the artifact folder name
```
Comment (freddydk) [inline Actions/Deliver/Deliver.ps1:171]: Both \ and / needs to be replaced with _ (for linux compatibility)

File: Actions/Deliver/Deliver.ps1 [hunk 2]
```diff
@@ -163,37 +168,42 @@ try {
...
+                # Get the folders holding the artifacts from all build modes
+                # Build modes are identified as a prefix to the artifact folder name
+                $artifactsFolders = @(Get-ChildItem -Path (Join-Path $baseFolder $("*" + $singleArtifactFilter)) -Directory)
+
+                if ($artifactsFolders.Count -gt 0) {
+                    $parameters[$artifactType.ToLower() + "Folders"] = $artifactsFolders.FullName
+                }
             }
+            
+            Write-Host "Calling custom script: $customScript with parameters: $($parameters | ConvertTo-Json -Depth 1)"
```
Comment (freddydk) [inline Actions/Deliver/Deliver.ps1:206]: We shouldn't print parameters here, as they contain secrets.
Yes, secrets are masked but still.

File: Actions/Deliver/Deliver.ps1 [hunk 3]
```diff
@@ -163,37 +168,42 @@ try {
...
-            else {
-                $parameters.dependenciesFolder = ""
+            $projectToFilter = $project.Replace('\','_') # this is the project name as it appears in the artifact folder name
+            
+            #Calculate the folders per artifact type
+            'Apps', 'TestApps', 'Dependencies' | ForEach-Object {
+                $artifactType = $_
+                $singleArtifactFilter = "$projectToFilter-$refname-$artifactType-*.*.*.*";
+
+                # Get the folder hoding the artifacts from the standard build
```
Comment (freddydk) [inline Actions/Deliver/Deliver.ps1:178]: hoding -> holding

File: Actions/Deliver/Deliver.ps1 [hunk 4]
```diff
@@ -163,37 +168,42 @@ try {
...
+
+                # Verify that there is only at most one artifact folder for the standard build
+                if ($artifactFolder.Count -gt 1) {
+                    $artifactFolder | Out-Host
+                    throw "Internal error - multiple $artifactType folders located"
+                }
+
+                # Add the artifact folder to the parameters
+                if ($artifactFolder.Count -ne 0) {
+                    $parameters[$artifactType.ToLower() + "Folder"] = $artifactFolder[0].FullName
```
Comment (freddydk) [inline Actions/Deliver/Deliver.ps1:194]: I always use ToLowerInvariant()
I know that there is no difference here

File: Actions/Deliver/Deliver.ps1 [hunk 5]
```diff
@@ -163,37 +168,42 @@ try {
...
+                }
+
+                # Get the folders holding the artifacts from all build modes
+                # Build modes are identified as a prefix to the artifact folder name
+                $artifactsFolders = @(Get-ChildItem -Path (Join-Path $baseFolder $("*" + $singleArtifactFilter)) -Directory)
```
Comment (freddydk) [inline Actions/Deliver/Deliver.ps1:199]: Might not work under linux (not sure) - but I will create a number of extra tests and run all tests under linux and windows - these will reveal that, so I am fine with this.

**Review by freddydk — CHANGES_REQUESTED** [review]

File: Actions/Deliver/Deliver.ps1 (hunk 1 above)
Comment (mazhelez) [reply Actions/Deliver/Deliver.ps1:171]:
```suggestion
            $projectToFilter = $project.Replace('\','_').Replace('\','_') # this is the project name as it appears in the artifact folder name
```

File: Actions/Deliver/Deliver.ps1 (hunk 3 above)
Comment (mazhelez) [reply Actions/Deliver/Deliver.ps1:178]:
```suggestion
                # Get the folder holding the artifacts from the standard build
```

File: Actions/Deliver/Deliver.ps1 (hunk 4 above)
Comment (mazhelez) [reply Actions/Deliver/Deliver.ps1:194]:
```suggestion
                    $parameters[$artifactType.ToLowerInvariant() + "Folder"] = $artifactFolder[0].FullName
```

File: Actions/Deliver/Deliver.ps1 (hunk 2 above)
Comment (mazhelez) [inline Actions/Deliver/Deliver.ps1:206]:
```suggestion
            Write-Host "Calling custom script: $customScript"
```

Comment (mazhelez) [reply Actions/Deliver/Deliver.ps1:206]:
```suggestion
            Write-Host "Calling custom script: $customScript"
```

File: Actions/Deliver/Deliver.ps1 [hunk 6]
```diff
@@ -163,37 +168,42 @@ try {
...
+
+                if ($artifactsFolders.Count -gt 0) {
+                    $parameters[$artifactType.ToLower() + "Folders"] = $artifactsFolders.FullName
```
Comment (mazhelez) [inline Actions/Deliver/Deliver.ps1:202]: Add some docs to https://github.com/microsoft/AL-Go/blob/main/Scenarios/settings.md

File: Actions/RunPipeline/RunPipeline.ps1 [hunk 7]
```diff
@@ -340,7 +343,36 @@ try {
...
+            Write-Host "Adding Preprocessor symbols: $preprocessorsymbols"
+            $runAlPipelineParams["preprocessorsymbols"] += $preprocessorsymbols
+        }
+        'LCGTranslated' {
+            if (!$runAlPipelineParams.ContainsKey('features')) {
+                $runAlPipelineParams["features"] = @()
+            }
+            $runAlPipelineParams["features"] += "lcgtranslationfile"
+        }
+        'Translated' {
```
Comment (aholstrup1) [inline Actions/RunPipeline/RunPipeline.ps1:367]: Remove this.. When running with Translated (-> Default) buildmode we shouldn't add any parameters

Comment (mazhelez) [reply Actions/RunPipeline/RunPipeline.ps1:361]: _Default_ build mode doesn't add anything extra

File: Actions/RunPipeline/action.yaml [hunk 8]
```diff
@@ -29,6 +29,10 @@ inputs:
     description: Secrets from repository in compressed Json format
     required: false
     default: '{"insiderSasToken":"","licenseFileUrl":"","CodeSignCertificateUrl":"","CodeSignCertificatePw":"","KeyVaultCertificateUrl":"","KeyVaultCertificatePw":"","KeyVaultClientId":"","applicationInsightsConnectionString": ""}'
+  buildMode:
+    description: Build mode
```
[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/RunPipeline/action.yaml:33]: "Specifies a mode to use for the build steps. _Default_ mode builds the code as is."

File: Scenarios/settings.md [hunk 9]
```diff
@@ -134,6 +136,18 @@ $parameters.ProjectSettings | Out-Host
...
+| Parameter | Description | Example |
+| --------- | :--- | :--- |
+| `$parametes.project` | The current AL-Go project | Root/AllProjects/MyProject
+| `$parameters.projectsName` | The name of the current AL-Go project | Root_AllProjects_MyProject
+| `$parameters.type` | Type of delivery (CD or Release) | CD
+| `$parameters.appsFolder` | The folder that contain the output from the default build of the non-test apps in the AL-Go project | AllProjects_MyProject-main-Apps-1.0.0.0
+| `$parameters.testAppsFolder` | The folder that contain the output from the default build of the test apps in the AL-Go project | AllProjects_MyProject-main-TestApps-1.0.0.0
+| `$parameters.appsFolders` | The folders that contain the output from all builds (from different build modes) of the non-test apps in the AL-Go project | AllProjects_MyProject-main-Apps-1.0.0.0, AllProjects_MyProject-main-CleanApps-1.0.0.0
+| `$parameters.testAppsFolders` | The folders that contain the output from all builds (from different build modes) of the test apps in the AL-Go project | AllProjects_MyProject-main-TestApps-1.0.0.0, AllProjects_MyProject-main-CleanTestApps-1.0.0.0
+
```
Comment (freddydk) [inline Scenarios/settings.md:150]: There is also a type called Dependencies (not used in ALAppExtensions, but in other repos)

File: Templates/AppSource App/.github/workflows/CICD.yaml [hunk 10]
```diff
@@ -287,8 +288,9 @@ jobs:
     strategy:
       matrix:
         project: ${{ fromJson(needs.Initialization.outputs.projects) }}
+        buildMode: ${{ fromJson(needs.Initialization.outputs.buildModes) }}
       fail-fast: false
-    name: Build ${{ matrix.project }}
+    name: Build ${{ matrix.project }} - ${{ matrix.buildMode }}
```
[the remark this answers is not in the document]
Comment (mazhelez) [reply Templates/AppSource App/.github/workflows/CICD.yaml:293]: Yes.

**Review by freddydk — APPROVED** [review]

File: Scenarios/settings.md (hunk 9 above)
Comment (mazhelez) [reply Scenarios/settings.md:152]: Done.

**Pull request merged**